### PR TITLE
Local Scope entries in new Hierarchical Action List.

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

Make the local scope components available to the Component Browser through `component::List` (previously known as "Hierarchical Action List"). (See the [Local Scope Section in the Design Doc](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#local-scope-section) for more details.)

https://www.pivotaltracker.com/story/show/181869186

### Important Notes

 - Note: the PR description does not include a screencast due to the changes in this PR not having any visual effect at this moment. The result of this PR's changes would be only observed in the Component Browser, but the Component Browser is not merged yet. As described in the task's Acceptance Criteria, the changes in this PR are currently only covered by tests.
 - The `component::group::Data::new_empty_visible` constructor was renamed to `from_name_and_id` and changed to set the `visible` field to a default value. All known code paths appear to call the `update_sorting_and_visibility` method before checking the value of the `visible` field, so setting the value to `true` when creating the object does not seem needed.

[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

**Comment by akavel** [comment]
**NOTE:** Converting to Draft, as this PR will need non-trivial conflict resolution with #3510 (which is merge-pending with "CI: Ready for merge").

File: app/gui/src/controller/searcher/component/builder.rs [hunk 1]
```diff
@@ -65,15 +65,26 @@ impl ModuleGroups {
...
+    /// `local_scope_module_id` (if set) will be cloned into [`component::List::local_scope`].
+    pub fn new(local_scope_module_id: Option<component::Id>) -> Self {
+        Self {
+            all_components: default(),
+            module_groups: default(),
+            local_scope_module_id,
+            local_scope_entries: default(),
+            favorites: default(),
+        }
     }
```
Comment (wdanilo) [inline app/gui/src/controller/searcher/component/builder.rs]: 1. Leave `new` as it was, it should be just `default`. If `Default` is implemented, `new` should behave the same way.
2. Create `pub fn with_local_scope_module_id(self, id: component::Id) -> Self`.

Use it as `List::new().with_local_scope_module_id(id)`.

File: app/gui/src/controller/searcher/component/builder.rs [hunk 2]
```diff
@@ -153,6 +172,8 @@ impl List {
                 flattened.update_sorting_and_visibility("");
             }
         }
+        let mut local_scope_entries = self.local_scope_entries;
+        component::group::sort(&mut local_scope_entries, "");
```
Comment (wdanilo) [inline app/gui/src/controller/searcher/component/builder.rs]: This is a strange API. When reading `group::sort(something, "")` the `""` is totally unexpected. What is it?

File: app/gui/src/controller/searcher/component/group.rs [hunk 3]
```diff
@@ -132,6 +117,29 @@ impl Group {
 
 
 
+/// ============
+/// === Sort ===
+/// ============
+
+/// Sort the components by [`Component::match_info`] if pattern is non-empty. Otherwise, sort
+/// non-module components alphabetically followed by modules alphabetically.
```
Comment (wdanilo) [inline app/gui/src/controller/searcher/component/group.rs]: If `pattern` is `""` then the components are sorted by a `match_info` - what is match info? I dont see it passed as an argument to this function.

File: app/gui/src/controller/searcher/component/builder.rs (hunk 2 above)
Comment (wdanilo) [reply app/gui/src/controller/searcher/component/builder.rs]: Edit, after reading the rest of the code, this should be called `sort_by_pattern` or something like that.

**Review by wdanilo — APPROVED** [review]
I did not check logic, so review from someone knowing the logic is required as well.

File: app/gui/src/controller/searcher.rs [hunk 4]
```diff
@@ -1169,13 +1171,22 @@ impl Searcher {
...
     builder
 }
 
+fn graph_module_qualified_name(
+    graph: &controller::ExecutedGraph,
+    project: &dyn model::project::API,
+) -> QualifiedName {
+    graph.graph().module.path().qualified_module_name(project.qualified_name())
+}
+
```
Comment (farmaazon) [inline app/gui/src/controller/searcher.rs]: This should be a method of (executed) graph controller.

File: app/gui/src/controller/searcher/component.rs [hunk 5]
```diff
@@ -128,6 +128,9 @@ pub struct List {
     top_modules_flattened: group::AlphabeticalList,
     module_groups:         Rc<HashMap<Id, ModuleGroups>>,
     filtered:              Rc<Cell<bool>>,
+    /// Components to display in the "Local Scope" section of the [Component
+    /// Browser](crate::controller::Searcher).
+    pub local_scope:       Rc<RefCell<Vec<Component>>>,
```
Comment (farmaazon) [inline app/gui/src/controller/searcher/component.rs]: Why is. this not just a component::Group?

File: app/gui/src/controller/searcher/component/group.rs (hunk 3 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher/component/group.rs]: > what is match info? I dont see it passed as an argument to this function.

MatchInfo is a field of Component.

File: app/gui/src/controller/searcher/component/group.rs [hunk 6]
```diff
@@ -132,6 +117,29 @@ impl Group {
...
+        components.sort_by(|a, b| {
+            let cmp_can_be_entered = a.can_be_entered().cmp(&b.can_be_entered());
+            cmp_can_be_entered.then_with(|| a.label().cmp(b.label()))
+        });
+    } else {
+        let cmp_match_info =
+            |a: &Component, b: &Component| a.match_info.borrow().cmp(&*b.match_info.borrow());
+        components.sort_by(|a, b| cmp_match_info(a, b).reverse());
+    }
+}
```
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/group.rs]: Now, I think that passing "pattern" here makes no sense - it's enough to pass an "is_filtering" bool instead.

**Review by farmaazon — APPROVED** [review]

File: app/gui/src/controller/searcher/component/builder.rs (hunk 2 above)
Comment (akavel) [reply app/gui/src/controller/searcher/component/builder.rs]: After addressing @farmaazon's comments, the `group::sort` function is no longer present. Additionally, I changed the interface of the very similar preexisting `Group::update_sorting_and_visibility` method, replacing its string argument with an enum argument.

File: app/gui/src/controller/searcher/component/group.rs (hunk 6 above)
Comment (akavel) [reply app/gui/src/controller/searcher/component/group.rs]: I changed the code to pass an enum instead, in order for it to be readable.

**Pull request merged**